Exceptions.md

```diff
@@ -0,0 +1,199 @@
+# Exceptions
+
+## Preliminaries
+
+By `Result<T>`, we mean the following datatype:
+
+```
+datatype Result<T> =
+| Success(value: T)
+| Failure(error: string)
+```
+
+For simplicity, we use "C#" to refer to any target language into which Dafny compiles.
+
+## Motivations
+
+### More readable Dafny code
+
+At the moment, we're writing a lot of code of the following form:
+
+```
+var res1: Result<T1> := operation1(args1);
+if res1.Failure? {
+	return Failure(res1.error);
+}
+var value1: T1 := res1.get;
+
+var res2: Result<T2> := operation2(args2);
+if res2.Failure? {
+	return Failure(res2.error);
+}
+var value2: T2 := res2.get;
+
+...
+```
+
+It would be nicer to have more concise code, which could look like this:
+```
+var value1: T1 :- operation1(args1);
+var value2: T2 :- operation2(args2);
+...
+```
+
+### Less glue code to interoperate with other languages
+
+Many Dafny libraries will live in a "sandwich" between two layers of C# code (or any other language Dafny compiles to):
+* Code written in C# by clients who find it cool to use verified software, but don't want to write Dafny themselves
+* Code written in Dafny, compiled to C#
+* Primitives written in C# which cannot be written in Dafny (eg networking, file system access, etc)
+
+Both the layer above and the layer below the Dafny layer use exceptions, and at the moment, we have to wrap the primitives to turn exceptions into something Dafny understands, and then wrap the Dafny-generated C# code into exception-producing code, because clients don't want to deal with types such as `Result` or `Option`, they want exceptions.
+It would be good to get rid of these two extra layers of wrappers.
+
+### Don't swallow exceptions if you can't handle them
+
+If an exception is thrown in the primitives layer, it should be possible to let it bubble up through the layers all the way up to the client code, because in many situations, only the client code knows how to deal with the exception appropriately.
+For instance, if the exception is a network failure, and we're in an application which has an offline mode, we don't want the lower layers to swallow the exception and return some dummy value, or terminate the program, but the exception should be passed on to the client app, which could just say, "ok so let's switch to offline mode".
+
+### Printing stack traces
+
+While a client is debugging their app, it would be useful for them to get a stack trace in case of unhandled exceptions which goes all the way through the stack, and is not interrupted by the Dafny layer.
+
+
+## Implementation
+
+The type `Result<T>` is hardcoded into Dafny, the same way as `seq`, `map`, `int`, etc.
```

Comment:
I would be happier to make `Result<T>` something more programmable. It could be a datatype with a recognizable form (for example, with 2 constructors), or with an identifying attribute (for example, `{:result}`), or with a type member with a particular name (for example, the datatype may contain a compilable predicate `Propagate` and a compilable function `Rewrap`). (Dafny currently does not allow datatypes have to user-defined members, but Dafny soon will.)